beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -1,84 +1,303 @@
+
+This approach aligns with the atomic nature of checkpoints and provides a more granular level of control compared to action-level rollbacks.
+
+## Example Scenario
+
+Consider a GitHub repository creation template with the following checkpoints:
+
+1. Create repository
+2. Setup branch protection rules
+3. Configure team access
+4. Create initial pull request
+
+If the template fails during the "Create initial pull request" step, the rollback process would:
+
+1. Attempt to rollback "Configure team access" by removing configured teams
+2. Attempt to rollback "Setup branch protection rules" by removing the rules
+3. Attempt to rollback "Create repository" by deleting the repository
+
+Each checkpoint's rollback function handles its specific resource cleanup, ensuring atomic operations with clear boundaries.
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Backstage UI
+    participant NunjucksWorkflowRunner
+    participant Action as "GitHub Action"
+    participant CP1 as "Checkpoint 1<br/>(Create Repo)"
+    participant CP2 as "Checkpoint 2<br/>(Branch Rules)"
+    participant CP3 as "Checkpoint 3<br/>(Team Access)"
+    participant CP4 as "Checkpoint 4<br/>(Create PR)"
+
+    User->>Backstage UI: Start Template Execution
+    Backstage UI->>NunjucksWorkflowRunner: Execute Template
+
+    Note over NunjucksWorkflowRunner: Normal Execution Flow
 
-- when user manually decide to perform this action.
-- task has to be recovered and TaskRecoverStrategy set to 'rollback'
+    NunjucksWorkflowRunner->>Action: Execute GitHub Action
+    Action->>CP1: Process (success)
+    CP1-->>Action: Complete with data
+    Action->>CP2: Process (success)
+    CP2-->>Action: Complete with data
+    Action->>CP3: Process (success)
+    CP3-->>Action: Complete with data
+    Action->>CP4: Process (fails)
+    CP4--xAction: Error
+    Action--xNunjucksWorkflowRunner: Action Failed
+
+    Note over NunjucksWorkflowRunner: Rollback Flow (LIFO order by checkpoint)
+
+    NunjucksWorkflowRunner->>CP3: Rollback (Team Access)
+    CP3-->>NunjucksWorkflowRunner: Rollback Complete
+    NunjucksWorkflowRunner->>CP2: Rollback (Branch Rules)
+    CP2-->>NunjucksWorkflowRunner: Rollback Complete
+    NunjucksWorkflowRunner->>CP1: Rollback (Repository)
+    CP1-->>NunjucksWorkflowRunner: Rollback Complete
+
+    NunjucksWorkflowRunner-->>Backstage UI: Template Execution Failed
+    Backstage UI-->>User: Display Error & Rollback Status
+```
 
 ## Design Details
 
-We are going to introduce an extra function in action API which will look like:
+### Extending Checkpoint API for Rollback
+
+The checkpoint API will be extended to include an optional rollback function that can utilize the data saved by the checkpoint:
+
+```typescript
+/**
+ * Extended Checkpoint API with rollback support
+ */
+async checkpoint<T>(
+  key: string,
+  fn: () => Promise<T>,
+  rollback?: (data: T) => Promise<void>
```

Comment:
I think it might make sense to wrap this up in an object too

```ts
rollback?: (opts: { data: T }) => Promise<void>
```

Makes it easier to evolve moving forward without breaking the API.

Reply:
Updated in https://github.com/backstage/backstage/pull/29196/commits/24c7b7093bb2ddb3fd5a7f03533e5ebd34d72351